### PR TITLE
pkg/k8sutil: fix annotation propagation

### LINKED ISSUE #4392: spec.podMetadata.annotations changes not applied starting with v0.50.0
**What happened?**

Starting with v0.50.0 when the `Prometheus` resource has an updated value for `spec.podMetadata.annotations` the changes are not applied to the StatefulSet. Prior versions updated as expected.

**Did you expect to see something different?**

Expected the StatefulSet to have the new annotations applied to the Pods.

**How to reproduce it (as minimally and precisely as possible)**:

1. Create a Prometheus resource without a value for `spec.podMetadata.annotations`
2. Add a value to `spec.podMetadata.annotations`

and

1. Create a Prometheus resource with a value for `spec.podMetadata.annotations`
2. Modify the value or remove the value from `spec.podMetadata.annotations`

**Environment**

* Prometheus Operator version:

0.50.0, 0.51.1, 0.51.2, v0.52.0

* Kubernetes version information:

Shouldn't be relevant but I tested with 1.19.x & 1.20.x clusters

**Anything else we need to know?**:

I believe this was introduced via https://github.com/prometheus-operator/prometheus-operator/pull/4185

### PR DESCRIPTION
**Pull request by fpetkovski**
## Description

A bug was introduced in 0.50.0 where annotations set in the Prometheus
podMetadata spec were not properly propagated to statefulset pods.

This commit fixes the bug and adds additional unit tests to verify the
functionality works as expected.

Fixes https://github.com/prometheus-operator/prometheus-operator/issues/4392

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [x] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Propagate annotations set in spec.podMetadata to the statefulset pods. Fixes #4392.
```

### CODE REVIEW

File: pkg/k8sutil/k8sutil.go [hunk 1]
```diff
@@ -254,12 +254,16 @@ func mergeMaps(new map[string]string, old map[string]string) map[string]string {
 }
 
 func mergeKubectlAnnotations(new *metav1.ObjectMeta, old metav1.ObjectMeta) {
-	new.SetAnnotations(mergeMapsByPrefix(new.Annotations, old.Annotations, "kubectl.kubernetes.io/"))
+	new.SetAnnotations(mergeMapsByPrefix(old.Annotations, new.Annotations, "kubectl.kubernetes.io/"))
```
Comment (simonpasquier) [inline pkg/k8sutil/k8sutil.go:257]: The `old` and `new` parameters are mixed up here compared to the function declaration and it's a bit confusing.

Comment (fpetkovski) [reply pkg/k8sutil/k8sutil.go]: Yes, that could be confusing. I swapped them here instead: https://github.com/prometheus-operator/prometheus-operator/pull/4422/files#diff-0128909c1627157f3265add5bfeacace324a45ac2aa7c81bf9e91e887edf67faR171.

We could also rename `old` and `new` into `from` and `to` to make it more clear what is being merged into what.

**Review by simonpasquier — APPROVED** [review]
lgtm

**Comment by fpetkovski** [comment]
Hm it looks like I might have introduced a regression, let me take another look

**Comment by fpetkovski** [comment]
Seems to have been a flake. 

@paulfantom any other comments or should we merge it?

**Review by paulfantom — APPROVED** [review]
lgtm

**Pull request merged**